File: elixir/lib/AoC01.ex
defmodule AoC01 do
  @doc ~S"""
  Maybe finds two numbers from a list that give `target` in sum and multiplies them

  ## Examples
      iex> AoC01.solve(20, [])
      {:nothing}
      iex> AoC01.solve(20, [1,1,2,3])
      {:nothing}
      iex> AoC01.solve(20,[10,2,3,3,4,4,4,10])
      {:just, 100}
      iex> AoC01.solves(2, 20, [10,10])
      {:just, 100}
      iex> AoC01.solves(3, 20, [10,5,5])
      {:just, 250}
      iex> AoC01.solves(10, 10, [1,1,1,1])
      {:nothing}
  """
  import Util
  def complements(target, x, xs), do: Enum.filter(xs, fn y -> y == target - x end)
  def solve(_ ,[]), do: nothing()
  def solve(target, [x|xs]), do: (cond do
     complements(target, x, xs) |> null -> solve(target, xs)
     true                              -> just(x * hd(complements(target, x, xs)))
  end)

  def solves(_, _, []), do: nothing()
  def solves(count, target, [x|xs]), do: (cond do
    count < 2  -> nothing()
    count == 2 -> solve(target, [x|xs])
    true       -> case solves((count - 1), (target - x), xs) do
      {:nothing} -> solves(count, target, xs)
      {:just, val} -> just(x*val)
    end
  end)
end

File: haskek/src/AoC01.hs
module AoC01 where

import Util

-- | Maybe finds two numbers from a list that give `target` in sum and multiplies them
aoc01 :: Int -> [Int] -> Maybe Int
aoc01 _ [] = Nothing
aoc01 target (x:xs) | null complements = aoc01 target xs
                    | otherwise        = Just $ x * head complements
  where complements = filter (==target-x) xs

-- | Maybe finds `count` numbers from a list that give `target` in sum and multiplies them
aoc01s :: Int -> Int -> [Int] -> Maybe Int
aoc01s _ _ [] = Nothing
aoc01s count target l@(x:xs) | count < 2  = Nothing
                             | count == 2 = aoc01 target l
                             | otherwise  = case aoc01s (pred count) (target-x) xs of
                                              Nothing -> aoc01s count target xs
                                              Just y  -> Just $ x * y

runAoC01 input = do
  let arrToSort = parseAsInts Nothing input
  print $ aoc01 2020 arrToSort

runAoC01s input = do
  let arrToSort = parseAsInts Nothing input
  print $ aoc01s 3 2020 arrToSort